src/hotspot/share/classfile/systemDictionaryShared.cpp

```diff
@@ -629,7 +646,79 @@ class UnregisteredClassesDuplicationChecker : StackObj {
   }
 };
 
-void SystemDictionaryShared::check_excluded_classes() {
+void SystemDictionaryShared::scan_constant_pool(InstanceKlass* k) {
+  if (CDSConfig::is_dumping_invokedynamic()) {
+    k->constants()->find_required_hidden_classes();
+  }
+}
+
+bool SystemDictionaryShared::should_hidden_class_be_archived(InstanceKlass* k) {
+  assert(k->is_hidden(), "sanity");
+  if (is_registered_lambda_proxy_class(k)) {
+    return true;
+  }
+
+  if (CDSConfig::is_dumping_invokedynamic()) {
+    DumpTimeClassInfo* info = _dumptime_table->get(k);
+    if (info != nullptr && info->is_required_hidden_class()) {
+      return true;
+    }
+  }
+
+  return false;
+}
+
+// Returns true if the class should be excluded. This can be called before
+// SystemDictionaryShared::find_all_archivable_classes().
+bool SystemDictionaryShared::check_for_exclusion(Klass* k) {
+  assert(CDSConfig::is_dumping_archive(), "sanity");
+
+  if (k->is_objArray_klass()) {
+    return check_for_exclusion(ObjArrayKlass::cast(k)->bottom_klass());
+  }
+
+  if (!k->is_instance_klass()) {
+    return false;
+  } else {
+    InstanceKlass* ik = InstanceKlass::cast(k);
+
+    if (SafepointSynchronize::is_at_safepoint()) {
```

Comment:
I renamed the public function to `should_be_excluded(InstanceKlass*)` to avoid confusion with the private function `check_for_exclusion(InstanceKlass*, DumpTimeClassInfo*)`. I also added code to make sure that exclusion has been checked for the class, even when in safepoint.